File: ios/Talise/Features/Rewards/RoundupCard.swift
import SwiftUI

/// Round-up & Save card — Phase 2 of the Rewards refresh.
///
/// Lets the user opt in to auto-saving a small percentage of every
/// outbound send (default 2%, configurable 1-10) to NAVI. Funds stay
/// in their wallet; they earn 5 pts per $1 swept.
///
/// State flow:
///   1. Card renders from `summary.roundup` + `summary.roundupSavedUsd`
///      (read on the parent's `load()`).
///   2. Toggling on/off or sliding the % POSTs to
///      `/api/rewards/roundup` and on success invokes `onChange()` —
///      the parent re-fetches the summary so the displayed values
///      come from the server, not local state. This keeps the
///      "saved via round-up" line and the toggle in lock-step with
///      DB truth.
///
/// The card sits between the lifetime-stats row and the earn-rules
/// card on RewardsView — see the `// ANCHOR: roundup-section` marker.
struct RoundupCard: View {
    /// Latest rewards summary from the parent. May be nil while the
    /// initial fetch is still in flight — the card renders in a
    /// disabled-skeleton state until config lands.
    let summary: RewardsSummary?

    /// Called after a successful config POST so the parent can refetch
    /// `/api/referral/summary` and pick up the new `roundup` + lifetime
    /// numbers. Wrapped in `Task { await load() }` at the call site.
    let onChange: () -> Void

    @State private var pendingToggle: Bool? = nil
    @State private var pendingPercentage: Int? = nil
    @State private var saving = false
    @State private var error: String? = nil

    /// What the UI currently shows — `pendingX` overrides during the
    /// optimistic flip so the toggle / slider feel instant; falls back
    /// to the server-truth value once the response lands and onChange
    /// triggers a parent refetch.
    private var enabled: Bool {
        pendingToggle ?? summary?.roundup?.enabled ?? false
    }
    private var percentage: Int {
        pendingPercentage ?? summary?.roundup?.percentage ?? 2
    }
    private var savedUsd: Double {
        summary?.roundupSavedUsd ?? 0
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            header

            if enabled {
                Divider().background(Color.white.opacity(0.06))
                slider
                Divider().background(Color.white.opacity(0.06))
                savedLine
            }

            footer
        }
        .padding(18)
        .frame(maxWidth: .infinity, alignment: .leading)
        .taliseGlass(cornerRadius: 22)
        .opacity(enabled ? 1.0 : 0.92)
        .animation(.easeInOut(duration: 0.18), value: enabled)
    }

    // MARK: - Header (title + subtitle + toggle)

    private var header: some View {
        HStack(alignment: .top, spacing: 12) {
            VStack(alignment: .leading, spacing: 4) {
                MicroLabel(text: "Round-up & Save", color: TaliseColor.fgDim)
                    .kerning(1.5)
                Text(subtitleCopy)
                    .font(TaliseFont.body(13, weight: .light))
                    .foregroundStyle(TaliseColor.fg)
                    .fixedSize(horizontal: false, vertical: true)
            }
            Spacer(minLength: 8)
            // The Toggle drives an optimistic local flip + a backend
            // POST. SwiftUI's binding semantics make a custom Binding
            // the cleanest way to keep the visual instant while the
            // network round-trip completes.
            Toggle(
                "",
                isOn: Binding(
                    get: { enabled },
                    set: { newValue in
                        Task { await save(enabled: newValue, percentage: nil) }
                    }
                )
            )
            .labelsHidden()
            .tint(TaliseColor.accent)
            .disabled(saving)
        }
    }

    /// One-line subtitle that updates with the current %. When the
    /// toggle is off we still show the default 2% to telegraph what
    /// the user is opting INTO — same copy the prompt called out.
    private var subtitleCopy: String {
        "Auto-save \(percentage)% of every send to Navi at 5% APY"
    }

    // MARK: - Slider (% picker)

    private var slider: some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack {
                MicroLabel(text: "Save percentage", color: TaliseColor.fgDim)
                    .kerning(1.5)
                Spacer()
                Text("\(percentage)%")
                    .font(TaliseFont.mono(13, weight: .light))
                    .foregroundStyle(TaliseColor.accent)
            }
            // Step:1 keeps the slider on integer percents (the backend
            // clamps to 1..10 ints anyway). onEditingChanged fires the
            // POST only on release, so dragging doesn't spam the API.
            Slider(
                value: Binding(
                    get: { Double(percentage) },
                    set: { pendingPercentage = clamp(Int($0.rounded()), 1, 10) }
                ),
                in: 1...10,
                step: 1,
                onEditingChanged: { editing in
                    if !editing, let p = pendingPercentage {
                        Task { await save(enabled: nil, percentage: p) }
                    }
                }
            )
            .tint(TaliseColor.accent)
            .disabled(saving)
        }
    }

    // MARK: - Saved-via-roundup line

    private var savedLine: some View {
        HStack(alignment: .firstTextBaseline) {
            VStack(alignment: .leading, spacing: 4) {
                MicroLabel(text: "Saved via round-up", color: TaliseColor.fgDim)
                    .kerning(1.5)
                Text(TaliseFormat.local2(savedUsd))
                    .font(TaliseFont.heading(20, weight: .medium))
                    .kerning(-0.6)
                    .foregroundStyle(TaliseColor.accent)
                    .lineLimit(1)
                    .minimumScaleFactor(0.6)
            }
            Spacer()
            Image(systemName: "leaf.fill")
                .font(.system(size: 18, weight: .medium))
                .foregroundStyle(TaliseColor.accent.opacity(0.7))
        }
    }

    // MARK: - Footer (how-it-works)

    private var footer: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text("How it works")
                .font(TaliseFont.mono(10, weight: .light))
                .foregroundStyle(TaliseColor.fgMuted)
                .kerning(1.2)
            Text("Every time you send money, we sweep your chosen percentage to Navi at ~5% APY. Funds stay in your wallet — and you earn 5 pts per $1 saved.")
                .font(TaliseFont.body(11, weight: .light))
                .foregroundStyle(TaliseColor.fgDim)
                .fixedSize(horizontal: false, vertical: true)
            if let error {
                Text(error)
                    .font(TaliseFont.mono(10, weight: .light))
                    .foregroundStyle(TaliseColor.danger)
                    .padding(.top, 2)
            }
        }
    }

    // MARK: - Network

    /// POSTs the updated config to `/api/rewards/roundup`. Optimistic:
    /// the pending* state lets the toggle/slider feel instant; on
    /// success we clear pendings + invoke `onChange()` to refetch the
    /// summary; on failure we revert + surface the error inline.
    private func save(enabled: Bool?, percentage: Int?) async {
        if saving { return }
        saving = true
        error = nil
        defer { saving = false }

        // Stage the optimistic value so the UI reflects intent
        // immediately (the explicit assignment also covers the case
        // where the caller passed `enabled` but the toggle was driven
        // by a tap that already set `pendingToggle` to the new value).
        if let enabled { pendingToggle = enabled }
        if let percentage { pendingPercentage = percentage }

        struct Body: Encodable { let enabled: Bool?; let percentage: Int? }
        struct Resp: Decodable { let enabled: Bool; let percentage: Int; let savedUsd: Double }

        do {
            let _: Resp = try await APIClient.shared.post(
                "/api/rewards/roundup",
                body: Body(enabled: enabled, percentage: percentage)
            )
            // Drop the optimistic shadow — parent refetch will populate
            // `summary.roundup` from server truth.
            pendingToggle = nil
            pendingPercentage = nil
            onChange()
        } catch {
            // Revert the optimistic flip so the toggle doesn't lie
            // about a state the server didn't accept.
            pendingToggle = nil
            pendingPercentage = nil
            self.error = "Couldn't update — try again."
        }
    }

    private func clamp(_ n: Int, _ lo: Int, _ hi: Int) -> Int {
        max(lo, min(hi, n))
    }
}

File: android/app/src/main/java/io/talise/app/feature/rewards/RoundupCard.kt
package io.talise.app.feature.rewards

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Spa
import androidx.compose.material3.Icon
import androidx.compose.material3.Slider
import androidx.compose.material3.SliderDefaults
import androidx.compose.material3.Switch
import androidx.compose.material3.SwitchDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.ViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.compose.viewModel
import io.talise.app.ui.theme.TaliseColors
import io.talise.app.ui.theme.TaliseType
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Round-up & Save config state — mirrors iOS `RoundupCard`'s optimistic
 * `pendingToggle` / `pendingPercentage` shadows. The shadow is held (NOT
 * cleared on the POST response) until the parent's refetched summary carries
 * the new value — the reconciler drops it seamlessly so the toggle never
 * flickers back to a stale snapshot.
 */
class RoundupViewModel : ViewModel() {

    data class State(
        val pendingToggle: Boolean? = null,
        val pendingPercentage: Int? = null,
        val saving: Boolean = false,
        val error: String? = null,
    )

    private val _state = MutableStateFlow(State())
    val state: StateFlow<State> = _state.asStateFlow()

    /** Stamp the optimistic percentage while the slider drags. */
    fun stagePercentage(p: Int) {
        _state.value = _state.value.copy(pendingPercentage = p.coerceIn(1, 10))
    }

    /** Drop each optimistic shadow ONLY once the server summary caught up. */
    fun reconcile(serverEnabled: Boolean?, serverPercentage: Int?) {
        val s = _state.value
        var next = s
        if (s.pendingToggle != null && serverEnabled == s.pendingToggle) next = next.copy(pendingToggle = null)
        if (s.pendingPercentage != null && serverPercentage == s.pendingPercentage) next = next.copy(pendingPercentage = null)
        if (next != s) _state.value = next
    }

    /** POST the updated config; optimistic, reverting on failure (iOS `save`). */
    fun save(enabled: Boolean?, percentage: Int?, onChange: () -> Unit) {
        if (_state.value.saving) return
        _state.value = _state.value.copy(
            saving = true,
            error = null,
            pendingToggle = enabled ?: _state.value.pendingToggle,
            pendingPercentage = percentage ?: _state.value.pendingPercentage,
        )
        viewModelScope.launch {
            runCatching {
                rewardsApi.roundup(RoundupUpdateRequest(enabled = enabled, percentage = percentage))
            }.onSuccess { resp ->
                // Pin the shadow to the SERVER-CONFIRMED values; the reconciler
                // clears it once the refetched summary matches.
                _state.value = _state.value.copy(
                    saving = false,
                    pendingToggle = resp.enabled,
                    pendingPercentage = resp.percentage,
                )
                onChange()
            }.onFailure {
                // Revert the optimistic flip so the toggle doesn't lie.
                _state.value = _state.value.copy(
                    saving = false,
                    pendingToggle = null,
                    pendingPercentage = null,
                    error = "Couldn't update. Try again.",
                )
            }
        }
    }
}

/**
 * Round-up & Save card — iOS `RoundupCard`. Opt in to auto-saving a small
 * percentage of every outbound send (default 2%, configurable 1-10).
 * Renders from the parent's `summary`; `onChange` asks the parent to
 * refetch `/api/referral/summary`.
 */
@Composable
fun RoundupCard(
    summary: RewardsSummary?,
    onChange: () -> Unit,
    vm: RoundupViewModel = viewModel(),
) {
    val state by vm.state.collectAsStateWithLifecycle()

    val enabled = state.pendingToggle ?: summary?.roundup?.enabled ?: false
    val percentage = state.pendingPercentage ?: summary?.roundup?.percentage ?: 2
    val savedUsd = summary?.roundupSavedUsd ?: 0.0

    // Reconcile the optimistic shadows against the refetched summary.
    LaunchedEffect(summary?.roundup?.enabled, summary?.roundup?.percentage) {
        vm.reconcile(summary?.roundup?.enabled, summary?.roundup?.percentage)
    }

    Column(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(20.dp))
            .background(TaliseColors.surface, RoundedCornerShape(20.dp))
            .alpha(if (enabled) 1.0f else 0.92f)
            .padding(20.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        // Header — eyebrow + subtitle + toggle.
        Row(verticalAlignment = Alignment.Top, horizontalArrangement = Arrangement.spacedBy(12.dp)) {
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(4.dp)) {
                Text(
                    "ROUND-UP & SAVE",
                    style = TaliseType.mono(10.sp),
                    letterSpacing = 2.0.sp,
                    color = TaliseColors.fgMuted,
                )
                Text(
                    "Auto-save $percentage% of every send and earn on the saved balance",
                    style = TaliseType.body(13.sp, FontWeight.Light),
                    color = TaliseColors.fgMuted,
                )
            }
            Switch(
                checked = enabled,
                onCheckedChange = { newValue ->
                    vm.save(enabled = newValue, percentage = null, onChange = onChange)
                },
                enabled = !state.saving,
                colors = SwitchDefaults.colors(
                    checkedTrackColor = TaliseColors.accent,
                    checkedThumbColor = Color.White,
                ),
            )
        }

        if (enabled) {
            RowDivider(inset = 18.dp)

            // Slider (% picker) — POST only on release so dragging doesn't spam.
            Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        "SAVE PERCENTAGE",
                        style = TaliseType.mono(10.sp),
                        letterSpacing = 2.0.sp,
                        color = TaliseColors.fgMuted,
                    )
                    Spacer(Modifier.weight(1f))
                    // White, not green — a setting readout, not an earnings figure.
                    Text(
                        "$percentage%",
                        style = TaliseType.heading(22.sp, FontWeight.Medium),
                        letterSpacing = (-0.8).sp,
                        color = TaliseColors.fg,
                    )
                }
                Slider(
                    value = percentage.toFloat(),
                    onValueChange = { vm.stagePercentage(it.toInt()) },
                    onValueChangeFinished = {
                        state.pendingPercentage?.let { p ->
                            vm.save(enabled = null, percentage = p, onChange = onChange)
                        }
                    },
                    valueRange = 1f..10f,
                    steps = 8,
                    enabled = !state.saving,
                    colors = SliderDefaults.colors(
                        thumbColor = TaliseColors.accent,
                        activeTrackColor = TaliseColors.accent,
                        inactiveTrackColor = TaliseColors.surface2,
                    ),
                )
            }

            RowDivider(inset = 18.dp)

            // Saved-via-roundup line — the ONE green hero on this card.
            Row(verticalAlignment = Alignment.CenterVertically) {
                Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                    Text(
                        "SAVED VIA ROUND-UP",
                        style = TaliseType.mono(10.sp),
                        letterSpacing = 2.0.sp,
                        color = TaliseColors.fgMuted,
                    )
                    Text(
                        local2(savedUsd),
                        style = TaliseType.heading(20.sp, FontWeight.Medium),
                        letterSpacing = (-0.8).sp,
                        color = TaliseColors.accent,
                        maxLines = 1,
                    )
                }
                Spacer(Modifier.weight(1f))
                Icon(
                    Icons.Filled.Spa,
                    contentDescription = null,
                    tint = TaliseColors.fgMuted,
                    modifier = Modifier.size(18.dp),
                )
            }
        }

        // Footer — only show the explainer in the OFF (opt-in) state.
        if (!enabled) {
            Text(
                "Funds stay in your wallet and earn 5 pts per $1 saved.",
                style = TaliseType.body(12.sp, FontWeight.Light),
                color = TaliseColors.fgDim,
            )
        }
        state.error?.let { error ->
            Text(
                error,
                style = TaliseType.mono(10.sp),
                color = TaliseColors.danger,
                modifier = Modifier.padding(top = 2.dp),
            )
        }
    }
}
